File: spring/jpa-study/web-jpa/src/main/java/com/keunwon/jpashop/service/OrderService.java
package com.keunwon.jpashop.service;

import com.keunwon.jpashop.domain.Delivery;
import com.keunwon.jpashop.domain.Member;
import com.keunwon.jpashop.domain.Order;
import com.keunwon.jpashop.domain.OrderItem;
import com.keunwon.jpashop.domain.item.Item;
import com.keunwon.jpashop.repository.ItemRepository;
import com.keunwon.jpashop.repository.MemberRepositoryOld;
import com.keunwon.jpashop.repository.OrderRepository;
import com.keunwon.jpashop.repository.OrderSearch;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@RequiredArgsConstructor
@Transactional(readOnly = true)
@Service
public class OrderService {

    private final OrderRepository orderRepository;
    private final MemberRepositoryOld memberRepositoryOld;
    private final ItemRepository itemRepository;


    @Transactional
    public Long order(Long memberId, Long itemId, int count) {
        Member member = memberRepositoryOld.findOne(memberId);
        Item item = itemRepository.findOne(itemId);

        Delivery delivery = new Delivery();
        delivery.setAddress(member.getAddress());

        OrderItem orderItem = OrderItem.createOrderItem(item, item.getPrice(), count);
        Order order = Order.createOrder(member, delivery, orderItem);

        orderRepository.save(order);
        return order.getId();
    }

    @Transactional
    public void cancelOrder(Long orderId) {
        Order order = orderRepository.findOne(orderId);
        order.cancel();
    }


    public List<Order> findOrders(OrderSearch orderSearch) {
        return orderRepository.findAll(orderSearch);
    }
}

File: spring/Spring-JWT/domains/domain-food/src/main/kotlin/com/github/keunwon/food/service/order/OrderService.kt
package com.github.keunwon.food.service.order

import com.github.keunwon.food.domain.order.OrderDeliveredService
import com.github.keunwon.food.domain.order.OrderPayedService
import com.github.keunwon.food.domain.order.OrderRepository
import com.github.keunwon.food.domain.order.OrderValidator
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class OrderService(
    private val orderRepository: OrderRepository,
    private val orderValidator: OrderValidator,
    private val orderMapper: OrderMapper,
    private val orderDeliveredService: OrderDeliveredService,
    private val orderPayedService: OrderPayedService,
) {
    @Transactional
    fun place(cart: Cart) {
        val order = orderMapper.mapFrom(cart).apply {
            place(orderValidator)
        }
        orderRepository.save(order)
    }

    @Transactional
    fun payOrder(orderId: Long) {
        orderPayedService.payOrder(orderId)
    }
   
    @Transactional
    fun deliverOrder(orderId: Long) {
        orderDeliveredService.deliverOrder(orderId)
    }
}
